### PR TITLE
This enables project(... VERSION x.y.z.a ...) and DESCRIPTION etc.. usage (IDFGH-11310)

### PR DESCRIPTION
**Pull request by kohait00**
the current implementation of project.cmake
overrides the 

project() 

function with a macro which then calls the overridden function, but only provides the project name along with some predefined LANGUAGES..
in case someone wanted to specify additional LANGUAGES, DESCRIPTION etc... this is entirely ignored.
probably to keep things simple.

this fix enables the usage of all the parameters for project() including LANGUAGES

it splits the provided ARGV when there is LANGUAGES present, and eliminates the predefined languages that the project.cmake already wants to specify itself.

everything else is untouched.

this also has the benefit that the cmake own PROJECT_VERSION now is able to be specified via the project(VERSION x.y.z.a ...) and then can be used to initalize the PROJECT_VER, no need to specify the PROJECT_VER explicitly as a set(PROJECT_VER x.y.z.a)..

I admit that the solution is probly not optimal, feel free to improve on it.

### CODE REVIEW

**Review by igrr — CHANGES_REQUESTED** [review]
Thanks for the nice improvement @kohait00!

The code looks good to me overall, I have left a few cosmetic comments. My main concern is that the part about `LANGUAGES` seems to be a piece of fairly complex string/list manipulation code. I think we need to have some test coverage for it to prevent regressions.

Do you think you can add a few test cases to validate the behavior of this code?

I see we have two test cases for the version related logic,

https://github.com/espressif/esp-idf/blob/b4268c874a4cf8fcf7c0c4153cffb76ad2ddda4e/tools/test_build_system/test_git.py#L28

and, somewhat badly named,
https://github.com/espressif/esp-idf/blob/b4268c874a4cf8fcf7c0c4153cffb76ad2ddda4e/tools/test_build_system/test_kconfig.py#L129

Probably it makes sense to
- create a new test file, test_versions.py
- move the two existing tests there
- add one new test case about version from `project` call
- add another new test case (there or in another file) to validate the logic related to `LANGUAGE` argument handling

If you don't have time to work on adding test cases, please let us know anyway.

File: tools/cmake/project.cmake [hunk 1]
```diff
@@ -445,8 +447,50 @@ macro(project project_name)
...
+    # handle LANGUAGES duplicates in the definition of project ARGV coming from top CMakeFile.txt
+    # LANGUAGES usually to be last topic in ARGV list, so we simply kill all the parts we also specify,
+    # and append the remainder
+
+    set(ARGV_EXT "LANGUAGES;C;CXX;ASM")
+
+    set(LANGUAGES_POS -1)
+    string(FIND "${ARGV}" "LANGUAGES" LANGUAGES_POS)
+    if (LANGUAGES_POS GREATER 0)
+        # LANGUAGES definition from top level CMakeFile.txt exists
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake:459]: I refer to the main CMakeFile.txt of the project that contains the project() entry
* main CMakeFile.txt ?
* project CMakeFile.txt?
* anything you find best describing it..

File: tools/cmake/project.cmake [hunk 2]
```diff
@@ -445,8 +447,50 @@ macro(project project_name)
         endif()
     endif()
 
+    # handle LANGUAGES duplicates in the definition of project ARGV coming from top CMakeFile.txt
+    # LANGUAGES usually to be last topic in ARGV list, so we simply kill all the parts we also specify,
+    # and append the remainder
+
+    set(ARGV_EXT "LANGUAGES;C;CXX;ASM")
+
+    set(LANGUAGES_POS -1)
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake:456]: valid, will do

File: tools/cmake/project.cmake [hunk 3]
```diff
@@ -445,8 +447,50 @@ macro(project project_name)
...
+        # ;ASM is unique enough
+        string(REPLACE ";ASM" "" ARGV_SUB "${ARGV_SUB}")
+
+        #last, remove the LANGUAGES keyword, which we specify ourselves, leave ; , might already be gone
+        string(REPLACE "LANGUAGES;" "" ARGV_SUB "${ARGV_SUB}")
+        string(REPLACE "LANGUAGES" "" ARGV_SUB "${ARGV_SUB}")
+
+        set(ARGV_PROJECT "${ARGV_PREFIX};${ARGV_EXT};${ARGV_SUB}")
+    else()
+        # no LANGUAGES definition from top CMakeFile.txt
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake:488]: yeah, still learning CMake :)

**Comment by kohait00** [comment]
regarding the oetending of test cases that @igrr suggested, I could try but to be honest, I am a newby in python, barely read it.. It would take some time for sure.. But then again, I am also learning CMake now..

Probably easier for me would be to provide some project(xxx) string that can be used as a starting point..

My suggestion:
I provide review changes ASAP, including ducumentation and the test case ideas for VERSION, LANGUAGES etc
Then depending on the time frame that you guys have we can decide who would write the unit tests

**Comment by kohait00** [comment]
the promised test cases:

syntax reference
https://cmake.org/cmake/help/latest/command/project.html

# usual simple case may not break. Nothing more than that was possible before
project(hello_world)

# this should be possible but wont work, due to project.cmake internal LANGUAGES definition
project(hello_world CSharp C CXX ASM)

# bit more sophisticated with version, probably the future default
project(hello_world VERSION 2.4.3)

#even more complicated, order as in CMake documentation
project(hello_world VERSION 2.4.3 DESCRIPTION "Cool New Project" LANGUAGES C CXX ASM)

#order reversed, also in the languages definitions
project(hello_world VERSION 2.4.3.6 LANGUAGES CXX C ASM DESCRIPTION "Cool New Project" )

# to trigger the special case of C at the end
project(hello_world VERSION 2.4.3 DESCRIPTION "Cool New Project" LANGUAGES CXX ASM C )

# adding more to languages
project(hello_world VERSION 2.4.3 DESCRIPTION "Cool New Project" LANGUAGES CSharp C CXX ASM )

# duplicates
project(hello_world VERSION 2.4.3 DESCRIPTION "Cool New Project" LANGUAGES CSharp C CXX ASM C )

# full blown, strict docu conform order
project(hello_world VERSION 2.4.3 DESCRIPTION "Cool New Project" LANGUAGES CSharp C CXX ASM HOMEPAGE_URL "http://www.hello-world.com")

# full blown, arbitrary order
project(hello_world DESCRIPTION "Cool New Project" VERSION 2.4.3.6 LANGUAGES CSharp C CXX ASM HOMEPAGE_URL "http://www.hello-world.com")

# this schows that not all cases are caught, description will result in 'Cool new Project', the C at the end is gone
project(hello_world VERSION 2.4.3.6 LANGUAGES CXX C ASM DESCRIPTION "Cool New ProjectC" )

# this wont ever work,
project(hello_world VERSION 2.4.3 DESCRIPTION "Cool New Project" CSharp C CXX ASM C )

File: tools/cmake/project.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (igrr) [reply tools/cmake/project.cmake:459]: "top level CMakeLists.txt" is okay, you just have "CMakeFile.txt" written instead.

File: tools/cmake/project.cmake [hunk 4]
```diff
@@ -74,6 +74,8 @@ function(__project_get_revision var)
         if(EXISTS "${_project_path}/version.txt")
             file(STRINGS "${_project_path}/version.txt" PROJECT_VER)
             set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${_project_path}/version.txt")
+        elseif (NOT "${PROJECT_VERSION}" STREQUAL "")
+            set(PROJECT_VER ${PROJECT_VERSION})
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake:78]: added documentation

File: tools/cmake/project.cmake [hunk 5]
```diff
@@ -445,8 +447,50 @@ macro(project project_name)
...
+
+    set(ARGV_EXT "LANGUAGES;C;CXX;ASM")
+
+    set(LANGUAGES_POS -1)
+    string(FIND "${ARGV}" "LANGUAGES" LANGUAGES_POS)
+    if (LANGUAGES_POS GREATER 0)
+        # LANGUAGES definition from top level CMakeFile.txt exists
+        string(SUBSTRING "${ARGV}" 0 ${LANGUAGES_POS} ARGV_PREFIX)
+        string(SUBSTRING "${ARGV}" ${LANGUAGES_POS} -1 ARGV_SUB)
+        # ARGV_SUB is "LANGUAGES X Y Z and the rest", just remove what we specify
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake]: i discribed it differntly, now it hopefully more clear what happens

File: tools/cmake/project.cmake [hunk 6]
```diff
@@ -445,8 +447,50 @@ macro(project project_name)
...
+
+        # the C definition is tricky, as it also conflicts with CSharp, CXX etc..
+        # so remove the definit option (;C; covers most, ;C at the end covers a special case.
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake]: removed in the following commits

File: tools/cmake/project.cmake [hunk 7]
```diff
@@ -445,8 +447,50 @@ macro(project project_name)
...
+        # ;C theoretically could be a DESCRITION or the like too, but unlikly
```
[the remark this answers is not in the document]
Comment (kohait00) [reply tools/cmake/project.cmake]: corrected

**Comment by kohait00** [comment]
I had to force push because I forgot to correct the 'unlikly' and fixed it via amend..

so that's the only change there.. 
I did an amend on f0d8b3f (which I have seen you  already cleared) and cherry pick on the documentation commit.. sorry for the hassle

there are no other changes so far

**Comment by sudeep-mohanty** [comment]
@kohait00 Thanks for the suggestions and apologies for getting back to you late on this. I would be pulling in your changes but we might have to enable each `project()` argument gradually with some unit-tests for sanity so that we do not introduce any regression or breaking changes.

**Comment by sudeep-mohanty** [comment]
sha=db01a1219fac2db85563a8baf388cb50c2136073

**Comment by sudeep-mohanty** [comment]
sha=96e2a8a51c7c9613d0699e46e00c4ca160b5eace

**Comment by kohait00** [comment]
> @kohait00 Thanks for the suggestions and apologies for getting back to you late on this. I would be pulling in your changes but we might have to enable each `project()` argument gradually with some unit-tests for sanity so that we do not introduce any regression or breaking changes. Therefore, I would incorporate the addition of the `VERSION` argument first and then investigate the others subsequently.

I dont mind at all.. Do whatever is desired, since I currently can't provide additional testcases apart from what I tested locally.
If there is anything else I can do here to support, let me know..
Have fun.

**Comment by Alvin1Zhang** [comment]
Thanks for contribution again, changes have been merged with https://github.com/espressif/esp-idf/commit/9beda4ce480de087b29df8750db38455d1de8f51.

**Pull request closed without merging**